packages/insured-bridge-relayer/src/CrossDomainFinalizer.ts

```diff
@@ -0,0 +1,91 @@
+import winston from "winston";
+
+import { createEtherscanLinkMarkdown, createFormatFunction, PublicNetworks } from "@uma/common";
+import { getAbi } from "@uma/contracts-node";
+import { InsuredBridgeL1Client, InsuredBridgeL2Client, GasEstimator } from "@uma/financial-templates-lib";
+
+export enum RelaySubmitType {
+  Slow,
+  SpeedUp,
+  Instant,
+  Ignore,
+}
+
+export class CrossDomainFinalizer {
+  constructor(
+    readonly logger: winston.Logger,
+    readonly gasEstimator: GasEstimator,
+    readonly l1Client: InsuredBridgeL1Client,
+    readonly l2Client: InsuredBridgeL2Client,
+    readonly account: string
+  ) {}
+  async checkForBridgeableL2TokensAndBridge() {
+    this.logger.debug({ at: "AcrossRelayer#CrossDomainFinalizer", message: "Checking bridgeable L2 tokens" });
+
+    const whitelistedTokenEvents = await this.l1Client.bridgeAdmin.getPastEvents("WhitelistToken", {
+      fromBlock: 0,
+      toBlock: await this.l1Client.l1Web3.eth.getBlockNumber(),
+    });
+
+    // Extract the l2Tokens that have been whitelisted. The new Set syntax acts to remove any duplicates from the array.
+    // This would be the case if a l2 token was re - whitelisted after upgrading the bridgePool contract on L1.
+    const whitelistedL2Tokens = [...new Set(whitelistedTokenEvents.map((event) => event.returnValues.l2Token))];
+
+    // Check if any of the whitelisted l2Tokens are bridgeable. Do this in one parallel call. Returns an array of bool
+    // for each l2Token, describing if it can be bridged from L2->L1.
+    const canBridge = await Promise.all(
+      whitelistedL2Tokens.map((l2Token) => this.l2Client.bridgeDepositBox.methods.canBridge(l2Token).call())
+    );
+
+    // For each canBridge result, check if it is true. If so, then we can bridge the token.
+    const bridgeableL2Tokens = whitelistedL2Tokens.filter((l2Token, index) => canBridge[index]);
+
+    // Finally, iterate over the bridgeable l2Tokens and bridge them.
+    if (bridgeableL2Tokens.length == 0) {
+      this.logger.debug({ at: "AcrossRelayer#CrossDomainFinalizer", message: "No bridgeable L2 tokens" });
+      return;
+    }
+    let nonceCounter = await this.l2Client.l2Web3.eth.getTransactionCount(this.account);
+    for (const l2Token of bridgeableL2Tokens) {
+      // Track the account nonce and manually increment on each TX. We need to do this because the L2 transactions
+      // process quicker than the infura node updates and we need to avoid the nonce collision.
+      try {
```

Comment:
nit: put the logic inside this `try{}` block into an internal method `_bridgeL2Token` to make this function more readable and easier to maintain.